### PR TITLE
Separate init and resume methods

### PR DESCRIPTION
**Pull request by trowski**
Since the state of the fiber must be known before calling `resume` for the first time to know what arguments to pass to the method, it makes more sense for the methods to also be separate. This allows the arg-info of `resume` to enforce the single argument, rather than having to check it within the method.

See [our green-thread implementation](https://github.com/amphp/green-thread/blob/eb68b4301c87b97e3bb2e15985234c0afefe8f65/src/functions.php#L34), where this initial resume call would be init instead.

This again is more symmetrical with `Generator`, calling `Generator::current()` to get the initial yield value, then subsequently calling `send`.

### CODE REVIEW

**Comment by lvht** [comment]
The reason that `Generator` has a dedicate `current` method is that the `Generator` the [**Interator Interface**](http://php.net/manual/en/class.iterator.php).

File: fiber.c [hunk 1]
```diff
@@ -323,35 +323,60 @@ ZEND_METHOD(Fiber, __construct)
...
+/* {{{ proto mixed Fiber::init(vars...)
  * Resume and send a value to the fiber */
-ZEND_METHOD(Fiber, resume)
+ZEND_METHOD(Fiber, init)
 {
 	zval *params;
 	uint32_t param_count;
 	zend_fiber *fiber;
-
+
```
Comment (kelunik) [inline fiber.c]: Whitespace should be stripped.

**Comment by kelunik** [comment]
@lvht I prefer a separate method as well, as already expressed in the internals discussion. On the other hand, those could also be directly accepted by `new Fiber` instead, which directly primes the fiber.

At least the three level `if` makes code harder to read.

**Comment by trowski** [comment]
@lvht Right, but IMO that influenced the name `Generator::current()` rather than the existence of the method. You can call `Generator::send()` on a new generator, but it executes passed the first yield, then immediately sends the given value, stopping at the second yield. To get the value of the first yield, you must call `Generator::current()`, which must be a separate method because you have nothing to send.

We have a similar situation here: For the initial call, you must pass variadic list of arguments to get the initial yield from the fiber. On subsequent calls, only a single value must be passed. This should be separate methods in my opinion, as very different things are happening.

**Comment by kelunik** [comment]
On a second thought I think that priming directly is a bad idea as long as `Fiber::yield()` can't be used inside internal functions. With an event loop like Amp you could create a new fiber inside `array_map()` for example, but only start it inside `Loop::defer()`, which will be a userland call.

**Comment by uasan** [comment]
The best way :)

```PHP
async function handle($a, $b){}; // declare handle Fiber
$fiber = handle(1, 2);           // instance new Fiber and first call handle
$fiber->resume(3);               // subsequent calls 
```

**Comment by trowski** [comment]
@uasan I fully agree, and would like to see that if this is merged into core. Using async/await, there's no need for an `init` function. For now though I wanted to have the best extension API possible, then move to the next step.

File: fiber.c (hunk 1 above)
Comment (trowski) [reply fiber.c]: Fixed.

**Comment by kelunik** [comment]
@uasan My preferred API is actually a bit different:

```php
function foobar($a, $b) {
    /* ... */
}

$x = foobar(1, 2); // <-- just "continues" the current fiber
$y = async foobar(1, 2); // <-- creates a new fiber

// $y doesn't have any methods, resumption is handled by the core
// you can yield placeholders, which are basically like Amp's promise
```

**Comment by lvht** [comment]
please see https://nikic.github.io/2012/12/22/Cooperative-multitasking-using-coroutines-in-PHP.html

```php
function gen() {
    $ret = (yield 'yield1');
    var_dump($ret);
    $ret = (yield 'yield2');
    var_dump($ret);
}

$gen = gen();
var_dump($gen->current());    // string(6) "yield1"
var_dump($gen->send('ret1')); 
```

`Fiber` is different to `Generator`.

There is not a `new Generator` process. Call function with `yield` keyword will get a generator object. So call this function is equivalent to **new generator** and **send parameter**. The `current` method will trigger the execution.

For `Fiber`, we cannot determine whether or not a call will be a fiber, because we may `Fiber::yield` in the sub method call. So we need a way to indicate we are executing a fiber. So I choose `new Fiber($callable)`;

But why `async` keyword? Because I want make all callable can be run in a fiber with out modification.

But why send parameters when init the Fiber like `new Fiber($callable, $a, $b, $c)`? Because I think a fiber is something like process and can be reused. Please see the undocumented `Fiber::reset` API.

Finally, both Lua's coroutine and Ruby's Fiber do offer the same `resume` API. There is no need to introduce a dedicate `init` method.

**Comment by lvht** [comment]
The keyword `async` is useless and **misleading**. A fiber is not always related to the async io process.

@uasan The style of `async function handle($a, $b){}; // declare handle Fiber` is not acceptable, because all callable should be run inside in a fiber without any modification.

@kelunik The style of `$y = async foobar(1, 2); ` is no acceptable as well. Because it is just a grammar sugar and is equivalent to `$f = new Fiber('foo', 1, 2)`.

**Comment by kelunik** [comment]
@lvht It's sugar, indeed, doesn't mean it's unacceptable. Actually, it's not entirely sugar, as `new Fiber` can't call `[$this, 'foo']` if `'foo'` is private, while `async $this->foo()` could work perfectly. It also avoids using the ugly callable syntax, which makes it more IDE friendly.

**Comment by lvht** [comment]
@kelunik 
> as new Fiber can't call [$this, 'foo'] if 'foo' is private, while async $this->foo() could work perfectly. 

The `async $this->foo()` is equivalent to `new Fiber([$this, 'foo'])`. You cannot call a object private method out that class.

> It also avoids using the ugly callable syntax, which makes it more IDE friendly.

The callable is a basic feature of PHP, no matter we need to face the fact.

**Comment by lvht** [comment]
@kelunik 
> // you can yield placeholders, which are basically like Amp's promise

The promise is also a grammar sugar. Promise is needless if we had stackful coroutine.

**Comment by uasan** [comment]
> @uasan The style of async function handle($a, $b){}; // declare handle Fiber is not acceptable, because all callable should be run inside in a fiber without any modification.

Perhaps i do not understand something, for the PHP compiler, the keyword async, just an instruction to compile the function `async function handle($a, $b){ /* code handle */ };` to like this code:
```PHP
function handle(...$args):Fiber
{
   $fiber = new Fiber(function($a, $b){ /* code handle */ });
   $fiber->resume(...$args);
   return $fiber;
};
```
What's in it not acceptable?

**Comment by trowski** [comment]
@lvht `async` does not imply **non-blocking IO**, only that the routine may be paused and resumed using the `await` keyword. If merging this into core, I think `$fiber = async funcName(1, 2)` would be a much better way to create `Fiber` instances. This would also require another method, such as `Fiber::current()` to get the value from the first `await`.

Another direction to go as an extension would be `$fiber = new Fiber('foo', 1, 2);` plus a `Fiber::current()` method. This method might be useful to get the last yielded value in addition to calling `Fiber::resume()`, similar to `Generator::current()`, as we do [here](https://github.com/amphp/amp/blob/4a742beb59615f36ed998e2dc210e36576e44c44/lib/InvalidYieldError.php#L12).

**Comment by kelunik** [comment]
> The callable is a basic feature of PHP, no matter we need to face the fact.

The cases where it's really needed are fortunately not that common. Fibers would be pretty common if you're writing async programs. The `async` could apply special rules (= the same rules as a call without `async`), while a generic `callable` as in `new Fiber` shouldn't do that for consistency reasons.

> The promise is also a grammar sugar. Promise is needless if we had stackful coroutine.

How do you combine two fibers otherwise? Like:

```
$a = async $socket1->write("foobar");
$b = async $socket2->write("hello");

await [$a, $b];
```

What's `$a` and `$b` then?

**Comment by lvht** [comment]
You have dived into to `Generator` and nodejs. 

I have said what I need to say. And I will persistent the current simple, uniform, Ruby and Lua style API.

**Comment by kelunik** [comment]
@lvht There doesn't need to be a scheduler. Promises as placeholders aren't bound to a scheduler. There's no need to support that specific short syntax, I'm fine with supporting `await $placeholder`, the `all` combinator can be implemented in userland.

**Pull request closed without merging**